### PR TITLE
CRM-18526: Fix upgrade code for foreign keys from CRM-18345

### PR DESCRIPTION
**Pull request by JKingsnorth**
- [CRM-18526: CRM-18345 breaks upgrades](https://issues.civicrm.org/jira/browse/CRM-18526)
  - [CRM-18345: Deleting an email address deletes any mailing events related to it, and the contact](https://issues.civicrm.org/jira/browse/CRM-18345)

### CODE REVIEW

**Comment by mfb** [comment]
I am guessing that you don't need to use the DELIMITER command in the update sql, because queries are executed by PHP not by the mysql CLI tool?

**Comment by totten** [comment]
FWIW, *.mysql.tpl cannot be passed directly to mysql. (Our programmatic MySQL interface only allows you to send a single statement at a time... which is awesome in terms of limiting SQL-injection vulnerabilities.) However, t run full multi-line SQL files, Civi has to parse out the lines on its own. Consequently, subtleties like DELIMITER probably don't work. The code which does this is `CRM_Utils_File::sourceSQLFile()`.

You might either:
 * Move this upgrade logic to `CRM/Upgrade/Incremental/php/FourSeven.php` and use multiple calls to `CRM_Core_DAO::executeQuery()` (instead of using `4.7.5.mysql.tpl`/`sourceSQLFile()`)
 * Update the parser in `CRM_Utils_File::sourceSQLFile()` to support `DELIMITER.`

(IMHO, it's easier to move the logic to `FourSeven.php` so that you don't need to use that parser.)

**Comment by JKingsnorth** [comment]
Thanks for the pointers. I'll take a look at this today.

**Comment by JKingsnorth** [comment]
This is still work in progress, but I'm not sure how to 'test' this upgrade task on a local box. Do the upgrade tests cover it?

**Comment by JKingsnorth** [comment]
Test failure is unrelated, but QA is still required - particularly around how I've included the variables in the DROP FOREIGN KEY. I could validate them as strings as this didn't work (the statement isn't expecting quoted strings). Nevertheless, this feels like a better approach than messing about with creating procedures, so thanks for the feedback so far.

**Comment by eileenmcnaughton** [comment]
This isn't the first time this problem was solved - back in 4.2 they used code like

```
    foreach ($tables as $tableName => $fKey) {
      $foreignKeyExists = CRM_Core_DAO::checkConstraintExists($tableName, $fKey);
      if ($foreignKeyExists) {
        CRM_Core_DAO::executeQuery("ALTER TABLE {$tableName} DROP FOREIGN KEY {$fKey}", $params, TRUE, NULL, FALSE, FALSE);
        CRM_Core_DAO::executeQuery("ALTER TABLE {$tableName} DROP INDEX {$fKey}", $params, TRUE, NULL, FALSE, FALSE);
      }
```

**Comment by eileenmcnaughton** [comment]
Realistically we should always use code like that to drop FKs. Unfortunately copy & paste fights back! I recently added code to 'safe-add' indexes. I *think* I retrospectively changed other calls as part of a rear-guard action against copy & paste (although I may have only considered doing that).

**Comment by JKingsnorth** [comment]
Check constraint exists! I didn't see that despite having a look around :)
hurrah. I'll refractor to use that instead.

**Comment by eileenmcnaughton** [comment]
You might like to wrap it as safeDeleteForeignKey - but putting it on the same class makes sense

**Comment by JKingsnorth** [comment]
@eileenmcnaughton Or add safeDeleteForeignKey as a new utility in CRM_Core_DAO? Maybe that would help it not be rewritten every time we needed to use it?

**Comment by eileenmcnaughton** [comment]
Yep - sounds good!

**Comment by JKingsnorth** [comment]
Argh, looks like there's a bug in checkConstraintExists(). It's just looking for the constraint name, but it might also be used for a key. For example the CREATE TABLE shows:

  KEY `FK_civicrm_mailing_recipients_mailing_id` (`mailing_id`),
  KEY `FK_civicrm_mailing_recipients_contact_id` (`contact_id`),
  KEY `FK_civicrm_mailing_recipients_email_id` (`email_id`),
  KEY `FK_civicrm_mailing_recipients_phone_id` (`phone_id`),
  CONSTRAINT `FK_civicrm_mailing_recipients_contact_id` FOREIGN KEY (`contact_id`) REFERENCES `civicrm_contact` (`id`) ON DELETE CASCADE,
  CONSTRAINT `FK_civicrm_mailing_recipients_email_id` FOREIGN KEY (`email_id`) REFERENCES `civicrm_email` (`id`) ON DELETE SET NULL,
  CONSTRAINT `FK_civicrm_mailing_recipients_mailing_id` FOREIGN KEY (`mailing_id`) REFERENCES `civicrm_mailing` (`id`) ON DELETE CASCADE

and checkConstraintExists on FK_civicrm_mailing_recipients_phone_id returns true, because it's finding the KEY rather than the CONSTRAINT.

**Comment by JKingsnorth** [comment]
Correction, it's not a bug. It's just checking for constraints, and of course KEY is a kind of 'constraint' as well. Perhaps this is one for a follow-up issue, we could extend the checkConstraintExists to accept what 'type' of constraint to look for (or any type if no value is passed in). Not sure I'll be able to work on that as part of this patch though.

**Comment by JKingsnorth** [comment]
Jenkins retest this please

File: CRM/Upgrade/Incremental/php/FourSeven.php [hunk 1]
```diff
@@ -192,6 +192,16 @@ public function upgrade_4_7_4($rev) {
    *
    * @param string $rev
    */
+  public function upgrade_4_7_5($rev) {
+    $this->addTask(ts('Upgrade DB to %1: SQL', array(1 => $rev)), 'runSql', $rev);
+    $this->addTask('Upgrade mailing foreign key constraints', 'upgradeMailingFKs');
+  }
```
Comment (colemanw) [inline CRM/Upgrade/Incremental/php/FourSeven.php]: This ought to be in `function upgrade_4_7_8`

**Comment by JKingsnorth** [comment]
@colemanw Done

**Comment by colemanw** [comment]
Code looks good to me, and tests are passing. I think we're good to go :)

**Pull request merged**